serde-reflection/src/de.rs

```diff
@@ -391,55 +396,151 @@ impl<'de, 'a> de::Deserializer<'de> for Deserializer<'de, 'a> {
         self.tracer
             .registry
-            .entry(name.to_string())
+            .entry(enum_name.to_string())
             .unify(ContainerFormat::Enum(BTreeMap::new()))?;
-        let known_variants = match self.tracer.registry.get_mut(name) {
+        let known_variants = match self.tracer.registry.get_mut(enum_name) {
             Some(ContainerFormat::Enum(x)) => x,
             _ => unreachable!(),
         };
-        // If we have found all the variants OR if the enum is marked as
-        // incomplete already, pick the first index.
-        let index = if known_variants.len() == variants.len()
-            || self.tracer.incomplete_enums.contains(name)
-        {
-            0
-        } else {
-            let mut index = known_variants.len() as u32;
-            // Scan the range 0..=known_variants.len() downwards to find the next
-            // variant index to explore.
-            while known_variants.contains_key(&index) {
-                index -= 1;
+
+        // If the enum is marked as incomplete, just visit the first index
+        // because we presume it avoids recursion.
+        if self.tracer.incomplete_enums.contains_key(enum_name) {
+            return visitor.visit_enum(EnumDeserializer::new(
+                self.tracer,
+                self.samples,
+                VariantId::Index(0),
+                &mut VariantFormat::unknown(),
+            ));
+        }
+
+        // First visit each of the variants by name according to `variants`.
+        // Later revisit them by u32 index until an index matching each of the
+        // named variants has been determined.
+        let provisional_min = u32::MAX - (variants.len() - 1) as u32;
+        for (i, &variant_name) in variants.iter().enumerate() {
+            if !self
+                .tracer
+                .discriminants
+                .contains_key(&(enum_type_id, VariantId::Name(variant_name)))
+            {
+                // Insert into known_variants with a provisional index.
+                let provisional_index = provisional_min + i as u32;
+                let variant = known_variants
+                    .entry(provisional_index)
+                    .or_insert_with(|| Named {
+                        name: variant_name.to_owned(),
+                        value: VariantFormat::unknown(),
+                    });
+                self.tracer
+                    .incomplete_enums
+                    .insert(enum_name.into(), EnumProgress::NamedVariantsRemaining);
+                // Compute the discriminant and format for this variant.
+                let mut value = variant.value.clone();
+                let enum_value = visitor.visit_enum(EnumDeserializer::new(
+                    self.tracer,
+                    self.samples,
+                    VariantId::Name(variant_name),
+                    &mut value,
+                ))?;
+                let discriminant = Discriminant::of(&enum_value);
```

Comment:
> I don't suppose we could infer the `index` from the `discriminant`?

You can turn a discriminant into an integer value as follows, but that is not necessarily the same `index` that would deserialize to that variant.

```rust
#[derive(Debug, PartialEq)]
enum Discriminant {
    U8(u8),
    U16(u16),
    U32(u32),
    U64(u64),
    U128(u128),
    Usize(usize),
    I8(i8),
    I16(i16),
    I32(i32),
    I64(i64),
    I128(i128),
    Isize(isize),
}

fn get_discriminant<T>(value: &T) -> Discriminant {
    struct GetDiscriminant(Option<Discriminant>);

    impl std::hash::Hasher for GetDiscriminant {
        fn write_u8(&mut self, i: u8) {
            assert!(self.0.replace(Discriminant::U8(i)).is_none());
        }
        fn write_u16(&mut self, i: u16) {
            assert!(self.0.replace(Discriminant::U16(i)).is_none());
        }
        fn write_u32(&mut self, i: u32) {
            assert!(self.0.replace(Discriminant::U32(i)).is_none());
        }
        fn write_u64(&mut self, i: u64) {
            assert!(self.0.replace(Discriminant::U64(i)).is_none());
        }
        fn write_u128(&mut self, i: u128) {
            assert!(self.0.replace(Discriminant::U128(i)).is_none());
        }
        fn write_usize(&mut self, i: usize) {
            assert!(self.0.replace(Discriminant::Usize(i)).is_none());
        }
        fn write_i8(&mut self, i: i8) {
            assert!(self.0.replace(Discriminant::I8(i)).is_none());
        }
        fn write_i16(&mut self, i: i16) {
            assert!(self.0.replace(Discriminant::I16(i)).is_none());
        }
        fn write_i32(&mut self, i: i32) {
            assert!(self.0.replace(Discriminant::I32(i)).is_none());
        }
        fn write_i64(&mut self, i: i64) {
            assert!(self.0.replace(Discriminant::I64(i)).is_none());
        }
        fn write_i128(&mut self, i: i128) {
            assert!(self.0.replace(Discriminant::I128(i)).is_none());
        }
        fn write_isize(&mut self, i: isize) {
            assert!(self.0.replace(Discriminant::Isize(i)).is_none());
        }
        fn write(&mut self, _bytes: &[u8]) {
            unreachable!()
        }
        fn finish(&self) -> u64 {
            unreachable!()
        }
    }

    let mut hasher = GetDiscriminant(None);
    std::hash::Hash::hash(&std::mem::discriminant(value), &mut hasher);
    hasher.0.unwrap()
}

fn main() {
    enum Enum { First, Second }
    println!("{:?}", get_discriminant(&Enum::Second));
}
```